### ISSUE TITLE
[data grid] Editing a row after deleting another causes uncaught error

### ISSUE THREAD

**Issue by HossamHa** [mui/mui-x#5983]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

When one tries to edit a row after deleting a different row, there is an uncaught error where the deleted row is not found.
This will only cause a console error and the functionality works, however, when trying the same and focusing a field programmatically the site crashes instead.

Maybe there is a stale state in the rowModesModel?

### Expected behavior 🤔

One should be able to edit rows after deleting some in the grid

### Steps to reproduce 🕹

Link to live example: https://codesandbox.io/s/uqtpzu?file=/demo.tsx

Steps:
Version 1: Only console error
1. Delete a row
2. Try to edit a cell of another row

Version 2: Site crashes
1. Change line 112 to ->   setRowModesModel({ ...rowModesModel, [id]: { mode: GridRowModes.Edit, fieldToFocus: 'age' } });
2. Delete a row
3. Try to edit a row

### Your environment 🌎

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Comment by m4theushw** [comment]
When you click to edit the cell we need to publish an event to inform that the cell currently focused has lost focus. The crash happens because the cell that lost focus is from the row that was deleted, so calling `apiRef.current.getCellParams` on it is not possible. The following diff fixes the problem:

```diff
diff --git a/packages/grid/x-data-grid/src/hooks/features/focus/useGridFocus.ts b/packages/grid/x-data-grid/src/hooks/features/focus/useGridFocus.ts
index 86658eb85..b7a1f40f7 100644
--- a/packages/grid/x-data-grid/src/hooks/features/focus/useGridFocus.ts
+++ b/packages/grid/x-data-grid/src/hooks/features/focus/useGridFocus.ts
@@ -56,7 +56,7 @@ export const useGridFocus = (
         return;
       }
 
-      if (focusedCell) {
+      if (focusedCell && apiRef.current.getRow(focusedCell.id)) {
         // There's a focused cell but another cell was clicked
         // Publishes an event to notify that the focus was lost
         apiRef.current.publishEvent(
```

Do you want to submit a PR?

**Comment by HossamHa** [comment]
@chaiwattsw If you could and it's fine by @m4theushw then just go ahead!

**Comment by m4theushw** [comment]
Will be fixed by #6009
